libraries/AP_Motors/AP_MotorsPulsing_Heli.cpp

```diff
@@ -0,0 +1,219 @@
+        // sends output to motors when armed but not flying
+        rc_write_angle(AP_MOTORS_MOT_2, 0);
+        rc_write_angle(AP_MOTORS_MOT_3, 0);
+        set_actuator_with_slew(_actuator[AP_MOTORS_MOT_1], actuator_spin_up_to_ground_idle()); // spin up motors
+        set_actuator_with_slew(_actuator[AP_MOTORS_MOT_4], actuator_spin_up_to_ground_idle());
+        rc_write(AP_MOTORS_MOT_1, output_to_pwm(_actuator[AP_MOTORS_MOT_1]));
+        rc_write(AP_MOTORS_MOT_4, output_to_pwm(_actuator[AP_MOTORS_MOT_4]));
+        break;
+    case SpoolState::SPOOLING_UP:
+    case SpoolState::THROTTLE_UNLIMITED:
+    case SpoolState::SPOOLING_DOWN:
+        // set motor output based on thrust requests
+        rc_write_angle(AP_MOTORS_MOT_2, _pitch_action * AP_MOTORS_COAX_SERVO_INPUT_RANGE); // pitch
+        rc_write_angle(AP_MOTORS_MOT_3, _roll_action * AP_MOTORS_COAX_SERVO_INPUT_RANGE); // roll
+        set_actuator_with_slew(_actuator[AP_MOTORS_MOT_1], thr_lin.thrust_to_actuator(_rotor_thrust));
+        set_actuator_with_slew(_actuator[AP_MOTORS_MOT_4], thr_lin.thrust_to_actuator(_tail_thrust));
+        rc_write(AP_MOTORS_MOT_1, output_to_pwm(_actuator[AP_MOTORS_MOT_1]));
+        rc_write(AP_MOTORS_MOT_4, output_to_pwm(_actuator[AP_MOTORS_MOT_4]));
+        break;
+    }
+
+}
+
+// get_motor_mask - returns a bitmask of which outputs are being used for motors or servos (1 means being used)
+//  this can be used to ensure other pwm outputs (i.e. for servos) do not conflict
+uint32_t AP_MotorsPulsing_Heli::get_motor_mask()
+{
+    //Our mask is the main rotor, pitch, roll, and the tail rotor
+    uint32_t motor_mask =
+        1U << AP_MOTORS_MOT_1 |
+        1U << AP_MOTORS_MOT_2 |
+        1U << AP_MOTORS_MOT_3 |
+        1U << AP_MOTORS_MOT_4;
+    uint32_t mask = motor_mask_to_srv_channel_mask(motor_mask);
+
+    // add parent's mask
+    mask |= AP_MotorsMulticopter::get_motor_mask();
+
+    return mask;
+}
+
+void AP_MotorsPulsing_Heli::output_armed_stabilizing()
+{
+    float   roll_thrust;                // roll thrust input value, +/- 1.0
+    float   pitch_thrust;               // pitch thrust input value, +/- 1.0
+    float   yaw_thrust;                 // yaw thrust input value, +/- 1.0
+    float   throttle_thrust;            // throttle thrust input value, 0.0 - 1.0
+    float   throttle_avg_max;           // throttle thrust average maximum value, 0.0 - 1.0
+    float   rp_scale = 1.0f;           // this is used to scale the roll, pitch and yaw to fit within the motor limits
+
+    // gyro ff and yaw ff. Also need rotor height above COM
+    Vector3f gyro_latest = _ahrs_view->get_gyro_latest();
+    float velocity_0 = 0;
+
+    float rotor_yaw_ff = _rotor_yaw_ff * velocity_0 * velocity_0;
+    Vector3f blade_omega(gyro_latest.x, gyro_latest.y, gyro_latest.z-velocity_0);
+
+    Vector3f omega_cross = gyro_latest%blade_omega;
+
+    float gyro_x_ff = omega_cross.x * _gyro_ff_gain;
+    float gyro_y_ff = omega_cross.y * _gyro_ff_gain;
```

Comment:
Typically "control" stuff would not be done in the motors lib. Can you explain why this is needed.

Reply:
@IamPete1 in this case the single rotor aircraft with a stiff blade can have can have a lot of rotational inertia built up and that can cause the aircraft to perform an action 90 degrees off of the current rotational velocity. This accounts for that. With multirotor vehicles this isn't needed because the gyro forces generally cancel. Where would you recommend it go?

Reply:
This is not just a phase offset of controls tho? Its a feed forward from measured rates? I still don't understand.

Reply:
Correct. Basically with a stiff rotor if you the rotate the vehicle at a given rate it will have some gyroscopic precession style force based on that rate, similar to when you spin up a bicycle wheel and try to move it around. The force acts 90deg later, which results in the precession motion. If I don't add a little bit of correction here then the vehicle precesses. Introducing this factor accounts for that. Phase offset is done within the motor.

Reply:
Sorry, I still don't understand why you would use the measured rates from the gyro. If it was target rates from the control loops I could see maybe. Our heli expert @bnsgeyer might understand better than I do.

Reply:
That's a fair question. You made me look a bit deeper and it seems that in the merging, the collection of rotor RPM was removed from this section. The reason I used measured rate instead of target rate was because the gyroscopic force that;s happening is related to how the rotor is actually moving. I guess it's a bit more delayed so theoretically if I'm tracking set point properly then the feedforward could act on a more up to date data. Once this commit gets in do you think you could look again to see if it makes more sense?

Reply:
@vertiq-jordan i think you need to have a detailed brief on what exactly you are doing here.  I’m not familiar with any of this and am not going to waste my time trying to figure out what you are doing by pouring through what appears to be massive change. 
it would be better received if there was some background presentation describing the hardware being used and the set up on the aircraft. What is a pulsing motor?  Are you using the swashplate with servos connected to it?  Is it powered by a brushless motor with an ESC?  If is is just a new way to control the motor powering the rotor system, then why does there need to be a change to the motors class?
you got to lead us ( or at least me) by the hand if you want to get constructive feedback.

Reply:
@bnsgeyer Sorry about that. We'll take a look and try to make something. To be clear, this does *not* use a swashplate. It uses pulsing of the motor to control cyclic pitch of our rotor system which you can find here:
https://www.vertiq.co/up12

The reason we need the PR is so that we can send more than one control signal to the motor controller at a time because the actuator is controlling 3 degrees of freedom.

Reply:
@vertiq-luca is there a reason why this couldn’t be implemented as a swashplate that somehow feeds the motor controller. I agree with @IamPete1 that you should not be compensating for rotor dynamics if you are just implementing a new way to control the rotor. Unless there is some physical reason which requires this compensation to properly control the vehicle because of how you are controlling the rotor tilt. 
there’s no doubt that rate feedback at the rotor level to compensate for the rate coupling would be good but it has to be done more genetically to affect all applicable heli types.

Reply:
@bnsgeyer That makes sense. When I was originally implementing this I was doing it as a heli instead of extending off of multirotors, but I found that since this is a fixed pitch helicopter there wasn't an existing type that really made sense. There are a lot of compensations that don't make sense when I'm actually spinning up and slowing down the rotor for control of collective. So I brought it over to multirotor, but then you're right, the compensations aren't in the right place really. 

The coax was also originally going to just fly using AP_MotorsCoax, but it has compensations for air velocity going over the control surfaces which aren't necessary in this case.

Reply:
@vertiq-luca ok. I didn’t realize that it is variable speed fixed pitch main rotor. That complicates things. Need to think about that